src/main/java/jenkins/plugins/slack/webhook/CommandRouter.java

```diff
@@ -0,0 +1,125 @@
+package jenkins.plugins.slack.webhook;
+
+
+import java.util.List;
+import java.util.ArrayList;
+
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+
+import java.lang.reflect.Method;
+
+import jenkins.plugins.slack.webhook.exception.CommandRouterException;
+import jenkins.plugins.slack.webhook.exception.RouteNotFoundException;
+
+import jenkins.plugins.slack.webhook.model.SlackTextMessage;
+
+
+
+
+public class CommandRouter<T> {
+
+    public CommandRouter() { }
+
+    public List<Route> routes = new ArrayList<Route>();
+
+    public CommandRouter addRoute(String regex,
+        String command,
+        String commandDescription,
+        RouterCommand routerCommand) {
+
+        this.routes.add(new CommandRouter.Route(regex,
+            command,
+            commandDescription,
+            routerCommand));
+
+        return this;
+    }
+
+    public List<Route> getRoutes() {
+        return this.routes;
+    }
+
+    public T route(String command) throws CommandRouterException,
+        RouteNotFoundException {
+
+        T message = null;
+
+        for (Route pa : routes) {
+
+            Matcher matcher = pa.regex.matcher(command);
+
+            boolean matches = matcher.matches();
+
+            if (matches) {
+    
+                List<String> parameters = new ArrayList<String>();
+
```

Comment:
The logic from here on down is a little confusing to me. I can see what's happening, but it feels like it may be a little over-complicated/misleading. `group(0)` returns the whole string that was matched, not the first matched group. If there are no groups matched then you really just need the whole string, otherwise you need the matches from 1 up, rather than getting everything from 0 up and then removing the first one.

If you find this acceptable, I think the following is less confusing, and is more concise.

```java
                String[] parametersArray = null;

                if (matcher.groupCount() == 0) {
                    parametersArray = new String[] { command };
                } else {
                    parametersArray = new String[matcher.groupCount()];
                    
                    for (int i = 1; i <= matcher.groupCount(); i++) {
                        parametersArray[i-1] = matcher.group(i);
                    }
                }

                // try..catch etc.
                pa.routerCommand.execute(parametersArray);
```